### DISCUSSION TITLE
第3章发布

### DISCUSSION THREAD

**Discussion by kanition** [kanition/pbrtbook#32] [发布更新]
~翻译至第3章，施工线在266页。之前的章节做少量调整。~
~branch: main commit: d2d959a23de586453bec6a56a19fc6ee16190d5b~
~branch: main commit: 8673829f269f006e18588a329369569802827f56~

请移步最新发布😉
<hr /><em>This discussion was created from the release <a href='https://github.com/kanition/pbrtbook/releases/tag/v0.3.0.1'>第3章发布</a>.</em>

**Comment by whysoslow** [comment]
从[CG Cinematography](https://chrisbrejon.com/cg-cinematography/introduction/)来的。
围观大佬翻译经典

**Reply by kanition** [reply]
我是初学者:joy: 不知道后面学不学得懂

**Comment by Thousandyearsofwar** [comment]
第二章的法线那一节有一句：
Note that, in an unfortunate turn of terminology, normals are not necessarily normalized.
感觉翻译有点小问题：
我感觉应该是因为空间转换运算，插值等原因法线要normalized

**Reply by kanition** [reply]
非常感谢你的反馈！！！
首先需要说明的是，文中的法线和法向量有时候是同一个意思（但严格区分的话，一个是直线，一个是向量，原文只写normal，所以没有细加区分了）。所以下面的回复中法线和法向量就不细分了。

从后文来看，正如你所说的这些原因，pbrt程序中用的法线是需要规范化的。
可这里却又强调地写出 _not_ necessarily，加上了unfortunate的语气。
所以我猜作者想说，数学概念上讲，法线的定义只要求垂直于局部平面，没有要求规范化，
但在程序中因为计算方便的原因是需要规范化的，
所以在介绍完程序中需要规范化的法线后，作者在这里澄清：通常意义上的法线没有规范化要求。
不知道你对这个理解的看法如何？

附：《数学辞海》[1]对法线的定义：
![图片](https://user-images.githubusercontent.com/68691995/136758417-6cdfd8d5-f88d-44f5-8505-19d3cb190223.png)

[1]何思谦, 裘光明总. 数学辞海.第一卷[M]. 中国科学技术出版社, 2002.

**Reply by Thousandyearsofwar** [reply]
非常感谢解答😁

**Comment by Thousandyearsofwar** [comment]
2.7.1齐次坐标的章节有一句存在错误：
这样有三个si 以及一个零或一的四维向量称作点或向量的齐次(homogeneous)表示
应该是第四维w=1为代表点,w=0为代表向量

**Reply by kanition** [reply]
非常感谢你的细致阅读！原文这里似乎也对应错了，后面我会把“点或向量”改为“向量和点”。

**Reply by kanition** [reply]
欢迎你提出更多建议和意见。有人回复太开心了！

**Comment by xwc2021** [comment]
好奇的問一下，分解旋转矩阵，如果只是M = RS，為什麼不直接normailize就好了阿
因為感覺S只會讓3軸的長度改變，並不會影響3軸的正交
https://photos.app.goo.gl/KENx2SuV5DemPSHs5

**Reply by kanition** [reply]
我以前也想过这个问题，以下是我学习后的看法，你也可以进一步向原作者询问。

首先，如果只是M=RS且S只是一个对角正定矩阵，我觉得你这种方法是可行的而且更简单，直接规范化就算出来了。
但是我猜测实践时又是另外一回事了。最大的问题是，没人能保证需要分解的M只含有这两个分量，这个时候就会出问题。
事实上式2.9强调了S不仅仅是对角的缩放矩阵（scale），也可能是更一般的拉伸矩阵（stretch）（此外第三章还有个错切（shear）矩阵，我感觉它们很像）。这种情况下就会影响三轴正交性了，但迭代的方法依然奏效。它实际上是优先保证了R的标准正交性，然后把剩余的分量都统统扔给S了，而这时S是否是一个纯粹的对角矩阵就不一定了。

你对这个解释的看法如何？

**Reply by xwc2021** [reply]
那這邊可能要修改一下喔
https://photos.app.goo.gl/2FuZspcriFR9Rfpk7
因為上面是打只有旋轉和縮放 
對也 非正交真的要考慮比較多 感謝 😆

**Reply by xwc2021** [reply]
想再請教一下關於 M=RS
v 是向量
如果上式是右乘 Mv=RSv

S可能是具有Sx、Sy、Sz元素的對角矩陣 
假設R是shear矩陣，那它的3軸長度就不一定是1，而且彼此不一定正交
R又受到先前S的影響，可能是R的x軸變長Sx倍、y軸變Sy倍、z軸變Sz倍

這樣的話
由2.91、 2.92求出的的R，指的是抵消掉受S影響的R嗎？
S指的是還原回Sx、Sy、Sz元素的對角矩陣嗎？

但我有看到你上面提到「它实际上是优先保证了R的标准正交性，然后把剩余的分量都统统扔给S了，而这时S是否是一个纯粹的对角矩阵就不一定了。」這裡就是我不太懂的部分

**Reply by kanition** [reply]
我想你比较纠结的就是，如果预设了一对R和S，算法能否给出一模一样的结果。我想它并不保证这一点。

这里我觉得换个角度看问题可能更清楚些。问题表述为：已知M，请从中提取出一对满足条件的R和S，即R是旋转矩阵（也就是标准正交矩阵），而不是要一定要“恢复”出原来的R和S。否则我们也没法判断得到的结果是不是“还原”了，因为原书提到我们除了知道M外没有任何其他信息了。

而且我个人认为没有必要“还原”，因为这里分解M的需求来自动画插值功能，为了合理地插值，需要（1）先求M的各分量，（2）再分别插值，（3）最后合成回去。只要这三步正确合理，动画插值功能就是对的。而要注意的是，旋转插值公式利用了旋转矩阵的标准正交性质，所以分解出的R必须保证是标准正交阵。当你提到“R是shear矩阵……彼此不一定正交”时，它其实已经不满足我们通常理解的旋转矩阵的条件了，我们不应该把它“恢复”出来，即便恢复出来了用于插值也是错的，因为它包含了除旋转以外的变换方式。

最后，我试着给了一个例子来说明算法并不保证能恢复你所说的的带“错切”的R和S：
预设R是旋转（绕x和y轴旋转90度）和错切混合的矩阵（不是标准正交的），即
R=[0, 0, 1; 1, 0, 0; 0, 1, 0] * [1, 0, 0.5; 0, 1, 0.5; 0, 0, 1] = [0, 0, 1; 1, 0, 0.5; 0, 1, 0.5]

S是对角阵即S=diag(3,4,5)。

则M=RS=[0, 0, 5; 3, 0, 2.5; 0, 4, 2.5]。

代入算法的结果是
R=[-0.2892,   -0.2554,    0.9226;
    0.9567,   -0.0439,    0.2878;
   -0.0329,    0.9658,    0.2570]

S=[2.8701,   -0.1318,    0.8633;
   -0.1318,    3.8634,    1.0280;
    0.8633,    1.0280,    5.9748]
你可以验证R确实是标准正交阵，但和你预设的完全不一样了。

你认可这个解释吗？

**Reply by kanition** [reply]
此外感谢你的建议，式2.91和式2.92中对S的说明，我会补充注明S是广义上的缩放矩阵。

**Reply by xwc2021** [reply]
感謝你的解惑 😋
所以你的意思是假設在2個關鍵影格
分別由有M1 和M2

(1)可以分解成正交的 R1 R2
和比較特別的 S1和S2
（2）再分别插值
（3）最后合成回去

那為什麼不一開始在2個關鍵影格
分別儲存旋轉信息,縮放和斜切值？

原來這些內容 本文都有討論到
我應該先看完再參與討論的
感謝你的耐心解答還有熱心翻譯 以後我會多注意的 🙇‍♂️

**Reply by kanition** [reply]
为什么不分开存储其实我也不太清楚，也许是节约存储的原因？不过我想到一点，对于同一个合成的M，不同渲染器可能分解的顺序不一样，万一有个渲染器是按M=SR处理的，结果我们提供的是M=RS，可能就乱套了。但是提供M的话，大家的处理是一模一样的。以上都是我的猜测。

欢迎继续提问和给出建议。本书内容很长，有问题很正常，我也是初学者，一遍学习一遍翻译啦。

**Reply by xwc2021** [reply]
我也是，很開心能一起討論
但我想到1點，如果是M=SR和M=RS的區別，分解出來的R，事前不知道是用左乘還右乘，用R旋轉的方向還是會相反；shear也是會和想來想的不太一樣吧？

**Reply by xwc2021** [reply]
> 此外感谢你的建议，式2.91和式2.92中对S的说明，我会补充注明S是广义上的缩放矩阵。

原來在這頁已經有提到，他指的縮放不是「當前座標的縮放」，而是「在某些座標系下的通用縮放」
https://photos.app.goo.gl/Fr2tbMYCBA2ecsUv5
是我自己跳著看才會有這個疑惑啦 😆

「在某些座標系下的通用縮放」我自己的解讀是，如果我們要況著(x軸+y軸)/2的方向縮放(也就是沿著45度角方向縮放)
可以把頂點 
(1)先變換到沿著z軸旋轉45度座標系的local space -> R_to
(2)再況著y軸縮放2倍 -> Sy
(3)再變換回原來的空間 -> R_back
所以S是1個複合矩陣 [R_back] [Sy] [R_to]

然後M就變成了 
M = RS = [R] [R_back] [Sy] [R_to]

但是_斜切矩陣_能不用寫成[R_back] [Sy] [R_to]的形式我就不太確定了

**Reply by kanition** [reply]
没错就是这个意思，某些坐标系下的通用缩放就说明了可能沿着一般的仿射坐标系进行缩放，在原来的标准直角坐标系下看时，S相当于含有复合成分。斜切矩阵我目前还没深入学习，我猜想后面关于相机透视变换的部分会详细介绍的。目前我才翻译完第4.3节，并且需要消化一下，所以后续内容还不了解。
